torch_geometric/explain/algorithm/pg_explainer.py

```diff
@@ -0,0 +1,276 @@
+import logging
+import warnings
+from typing import Optional
+
+import torch
+import torch.nn.functional as F
+import tqdm
+from torch import Tensor, nn
+from torch_scatter import scatter_mean
+
+from torch_geometric.explain import Explanation
+from torch_geometric.explain.algorithm.utils import clear_masks, set_masks
+from torch_geometric.explain.config import MaskType, ModelMode, ModelTaskLevel
+from torch_geometric.utils import get_intermediate_messagepassing_embeddings
+
+from .base import ExplainerAlgorithm
+
+
+# TODO add example.
+class PGExplainer(ExplainerAlgorithm):
+    r"""The PGExplainer model from the `"Parameterized Explainer for Graph
+    Neural Network" <https://arxiv.org/abs/2011.04573>`_ paper. It
+    uses a neural network :obj:`explainer_model` , to predict which
+    edges are crucial to a GNNs node or graph prediction.
+
+    Args:
+        epochs (int, optional): The number of epochs to train.
+            (default: :obj:`30`).
+        lr (float, optional): The learning rate to apply.
+            (default: :obj:`0.003`).
+        log (bool, optional): If set to :obj:`False`, will not log any learning
+            progress. (default: :obj:`True`)
+        **kwargs (optional): Additional hyper-parameters to override default
+            settings in
+            :attr:`~torch_geometric.explain.algorithm.PGExplainer.coeffs`.
+    """
+
+    coeffs = {
+        'edge_size': 0.05,
+        'edge_ent': 1.0,
+        'temp': [5.0, 2.0],
+        'bias': 0
+    }
+
+    def __init__(self, epochs: int = 30, lr: float = 0.003, log: bool = True,
+                 **kwargs):
+        super().__init__(training_need=True)
+        self.epochs = epochs
+        self.lr = lr
+        self.log = log
+        self.coeffs.update(kwargs)
+
+    def supports(self) -> bool:
+        task_level = self.model_config.task_level
+        if task_level not in [ModelTaskLevel.node, ModelTaskLevel.graph]:
+            logging.error(f"Task level '{task_level.value}' not supported")
+            return False
+
+        edge_mask_type = self.explainer_config.edge_mask_type
+        if edge_mask_type not in [MaskType.object]:
+            logging.error(f"Edge mask type '{edge_mask_type.value}' not "
+                          f"supported")
+            return False
+
+        node_mask_type = self.explainer_config.node_mask_type
+        if node_mask_type is not None:
+            logging.error(
+                "Node mask not supported. Set 'node_mask_type' to 'None'")
+            return False
+
+        is_hetero = self._is_hetero
+        if is_hetero:
+            logging.error("Heterogeneous graphs not supported.")
+            return False
+
+        return True
+
+    def train_explainer(self, model: torch.nn.Module, x: Tensor,
+                        edge_index: Tensor, target: Tensor = None,
+                        index: Tensor = None, **kwargs):
```

Comment:
Just a quick question for the interface update: does the `PGExplainer` need to be trained for every input like the `GNNExplainer` or is the training done once and then the user can explain multiple inputs ? If we are in the first case, we could do the training in the `PGExplainer.forward` method ? 

From here it seems like the training has to be done for each input ($G$,$y$) we want to explain, but I'm not sure if I understand correctly.

Reply:
In `PGExplainer` the training is done once and the user can explain multiple inputs with that. So its the second case, which is also the reason might need this interface change. Another thing we could do is have the user provide the training data while initializing `PGExplainer`.